### PR TITLE
Adapt multidict C ext test implementation

### PR DESCRIPTION
**Pull request by bdraco**
## What do these changes do?

Adds conftest facade for handling the C ext

## Are there changes in behavior for the user?

no

## Related issue number
#18

### CODE REVIEW

File: pytest.ini [hunk 1]
```diff
@@ -85,6 +85,4 @@ norecursedirs =
 
 testpaths = tests/
 
-pythonpath = src
```
Comment (bdraco) [inline pytest.ini:88]: Should revert this from the `src` PR in a separate PR since this likely was not needed

File: src/propcache/_helpers_c.pyx [hunk 2]
```diff
@@ -1,5 +1,13 @@
...
-from types import GenericAlias
+import sys
+import types
+
+if sys.version_info >= (3, 9):
+    GenericAlias = types.GenericAlias
+else:
+
+    def GenericAlias(cls):
+        return cls
```
Comment (bdraco) [inline src/propcache/_helpers_c.pyx:10]: Should fix this in a separate PR

Comment (bdraco) [reply src/propcache/_helpers_c.pyx:10]: https://github.com/aio-libs/propcache/pull/26

File: pytest.ini (hunk 1 above)
Comment (bdraco) [reply pytest.ini:88]: https://github.com/aio-libs/propcache/pull/27

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+from functools import cached_property
+from importlib import import_module
+from sys import version_info as _version_info
+from types import ModuleType
+from typing import List, Type, Union
+
+import pytest
+
+C_EXT_MARK = pytest.mark.c_extension
+PY_38_AND_BELOW = _version_info < (3, 9)
```
Comment (webknjaz) [inline tests/conftest.py:12]: this compat var is probably no longer necessary

Comment (bdraco) [reply tests/conftest.py:12]: I was waiting until 3.8 moved to unsupported on https://devguide.python.org/versions/

File: setup.cfg [hunk 4]
```diff
@@ -72,6 +72,8 @@ include_package_data = True
 # (see notes for the asterisk/`*` meaning)
 * =
   *.so
+  *.exp
```
Comment (webknjaz) [inline setup.cfg:75]: what's this file type?

File: tests/conftest.py (hunk 3 above)
Comment (webknjaz) [reply tests/conftest.py:12]: Oh, right. I just figured since you were removing it from other projects, you'd want that here too.

File: tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+
+def pytest_addoption(
+    parser: pytest.Parser,
+    pluginmanager: pytest.PytestPluginManager,
+) -> None:
+    """Define a new ``--c-extensions`` flag.
+
+    This lets the callers deselect tests executed against the C-extension
+    version of the ``multidict`` implementation.
```
Comment (webknjaz) [inline tests/conftest.py:77]:
```suggestion
    version of the ``propcache`` implementation.
```

File: tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    pluginmanager: pytest.PytestPluginManager,
+) -> None:
+    """Define a new ``--c-extensions`` flag.
+
+    This lets the callers deselect tests executed against the C-extension
+    version of the ``propcache`` implementation.
+    """
+    del pluginmanager
+
+    action: Union[str, Type[argparse.Action]]
```
Comment (webknjaz) [inline tests/conftest.py:81]: could you use a better var name than an action?

File: setup.cfg (hunk 4 above)
Comment (bdraco) [reply setup.cfg:75]: I think its an intermediate file produced on windows.  We probably don't need it.   I've been experimenting to fix the windows builds.

I haven't been able to get the windows build to succeed to build the c-ext.  I didn't realize it was failing after the `src` move until this PR

Comment (bdraco) [reply setup.cfg:75]: Just realized windows produces `pyd` instead of `.so` so I'll fix that.

I haven't used a windows system with python in 15 years...

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        """Return a text representation of the pure-python attribute."""
+        return "pure-python" if self.is_pure_python else "c-extension"
+
+    @cached_property
+    def imported_module(self) -> ModuleType:
+        """Return a loaded importable containing a propcache variant."""
+        importable_module = "_helpers_py" if self.is_pure_python else "_helpers_c"
+        return import_module(f"propcache.{importable_module}")
+
+    def __str__(self):
```
Comment (webknjaz) [inline tests/conftest.py:41]:
```suggestion
    def __str__(self) -> str:
```

File: tests/conftest.py (hunk 6 above)
Comment (bdraco) [reply tests/conftest.py]: 0349148ac41673b33b8b9dc722f4791d53ca6439

File: MANIFEST.in [hunk 8]
```diff
@@ -14,8 +14,8 @@ graft requirements
...
 global-exclude *.pyc
 global-exclude *.cache
-exclude propcache/*.c
-exclude propcache/*.html
-exclude propcache/*.so
-exclude propcache/*.pyd
+exclude src/propcache/*.c
+exclude src/propcache/*.html
+exclude src/propcache/*.so
+exclude src/propcache/*.pyd
```
Comment (bdraco) [inline MANIFEST.in:20]: Should have been done in the `src` PR.  Should be moved to a separate PR

File: setup.cfg (hunk 4 above)
Comment (webknjaz) [reply setup.cfg]: Oh, right. And I think macOS has a dynlib.

FWIW, there's improvements to be made with migrating to PEP 621 and `setuptools-scm` but I don't want to prioritize this right now. Perhaps, this is something to target during the sprint...

File: MANIFEST.in (hunk 8 above)
Comment (webknjaz) [reply MANIFEST.in:20]: I think that `.pyd` shouldn't be excluded, though. We don't have it, but it's one of the Cython source types IIRC...

File: setup.cfg (hunk 4 above)
Comment (bdraco) [reply setup.cfg]: looks like macos is producing `.so`

_helpers_c.cpython-38-darwin.so

File: MANIFEST.in (hunk 8 above)
Comment (bdraco) [reply MANIFEST.in:20]: I'm not sure if `.so` or `.pyd` should be excluded actually.

`html` and `c` from cython definitely should be

Comment (bdraco) [reply MANIFEST.in:20]: https://github.com/aio-libs/propcache/pull/29

Comment (bdraco) [inline MANIFEST.in:20]:
```suggestion
exclude propcache/*.c
exclude propcache/*.html
exclude propcache/*.so
exclude propcache/*.pyd
```
